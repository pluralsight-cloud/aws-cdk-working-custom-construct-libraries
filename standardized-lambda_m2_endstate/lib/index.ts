import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

const allowedRuntimes = [
  lambda.Runtime.NODEJS_22_X,
  lambda.Runtime.NODEJS_20_X,
  lambda.Runtime.PYTHON_3_13,
  lambda.Runtime.PYTHON_3_12
];

export interface StandardizedLambdaProps {
  runtime: lambda.Runtime;
  handler: string;
  code: lambda.Code;
  role?: iam.IRole;
  managedPolicies?: iam.IManagedPolicy[];
}

export class StandardizedLambda extends Construct {
  public readonly function: lambda.Function;

  constructor(scope: Construct, id: string, props: StandardizedLambdaProps) {
    super(scope, id);

    if (!allowedRuntimes.includes(props.runtime)) {
      throw new Error(`Invalid runtime. Allowed runtimes: ${allowedRuntimes.map(r => r.name).join(', ')}`);
    }    

    const role = props.role ?? new iam.Role(this, 'LambdaExecutionRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
        ...(props.managedPolicies || []), // User-defined policies
      ],
    });

    this.function = new lambda.Function(this, 'LambdaFunction', {
      runtime: props.runtime,
      handler: props.handler,
      code: props.code,
      role: role,
    });    

  }
}
