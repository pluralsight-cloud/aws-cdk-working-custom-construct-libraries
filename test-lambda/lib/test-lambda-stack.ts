import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as stdlambda from '@elle-la/standardized-lambda';
import { Construct } from 'constructs';
// import * as sqs from 'aws-cdk-lib/aws-sqs';

export class TestLambdaStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new stdlambda.StandardizedLambda(this, 'LambdaWithS3Access', {
      runtime: lambda.Runtime.NODEJS_22_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('lambda'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonS3FullAccess'),
      ],
    });    

  }
}
