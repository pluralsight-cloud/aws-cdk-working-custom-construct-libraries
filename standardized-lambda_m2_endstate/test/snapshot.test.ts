import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Template } from 'aws-cdk-lib/assertions';
import { StandardizedLambda } from '../lib/index';

describe('StandardizedLambda Snapshot', () => {
  test('Matches snapshot with default role', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'SnapshotWithDefaultRole');

    new StandardizedLambda(stack, 'LambdaWithDefaultRole', {
      runtime: lambda.Runtime.NODEJS_22_X,
      handler: 'index.handler',
      code: lambda.Code.fromInline('exports.handler = async () => {};'),
    });

    const template = Template.fromStack(stack).toJSON();
    expect(template).toMatchSnapshot();
    

  });
});
