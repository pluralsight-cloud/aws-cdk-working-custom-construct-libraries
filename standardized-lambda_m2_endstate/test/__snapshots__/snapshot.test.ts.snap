// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardizedLambda Snapshot Matches snapshot with default role 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaWithDefaultRoleLambdaExecutionRoleC97BBC5D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaWithDefaultRoleLambdaFunction9A204F5F": {
      "DependsOn": [
        "LambdaWithDefaultRoleLambdaExecutionRoleC97BBC5D",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LambdaWithDefaultRoleLambdaExecutionRoleC97BBC5D",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
